### ISSUE TITLE
readable classnames

### ISSUE THREAD

**Issue by threepointone** [threepointone/glamor#238]
I think I've figured out how to do 'readable' classnames. a quick example - 

```jsx
css({ 
  label: 'customStyle',
  color: 'red', 
  ':hover': {
    color: 'blue',
    fontSize: 20
  }
}) 
```
becomes ...
```
css-customstyle_color-red_hover-color-blue_fontsize-20
```
i.e, `[label]-[styles]`
another variant would be `[label]-[hash]`

some notes - 
- we can revert to hashed classnames in prod, keeping it tight/efficient. 
- label(s) accumulate at the beginning of the class, for quick readability 

open to feedback/questions/ideas.

**Comment by kwelch** [comment]
How do you feel about a second parameters rather than mixed with the styles? This would define purpose more clearly.

**Comment by threepointone** [comment]
Don't want to change/break current api. Making a wrapper shouldn't be hard tho.

**Comment by kentcdodds** [comment]
So long as we keep a standard prefix so I can programmatically determine that a class name is from glamor, I'm good. 👍

**Comment by donaldpipowitch** [comment]
I dunno. I have no big opinion on this. I never needed that for now and I could imagine that this can become quite unreadable with many styles...?

On Twitter you said that this is the number one request. Can you explain what the people would like to achieve with this? Better debuggability? Maybe this could be achieved in a different way (Source Maps for inline styles, dev tools, ...)

**Comment by kentcdodds** [comment]
> I could imagine that this can become quite unreadable with many styles

Agreed. I think it should just be `css-[label]-[hash]` and it could be that in prod for all I care (could write a babel plugin to remove the `label` as well for people who really care about bytes).

**Comment by threepointone** [comment]
@donaldpipowitch Can't do sourcemaps with dynamic code, investigated it a bunch. Can cap at X characters to prevent extra long classnames. 

Heck, I could try to make this userspace configurable, so you could drop in an emoji classname generator if you'd like. :D

**Comment by kentcdodds** [comment]
> Can't do sourcemaps with dynamic code

I'm very interested to see what sourcemaps would look like with glamorous! That'd be super cool!

If you cap at X characters, I could see collisions happening fairly easily though 😖 Maybe require a `label`? That seems kinda messy though... :-(

**Comment by threepointone** [comment]
Can attach the hash to truncated classnames

**Comment by kentcdodds** [comment]
But it's generated at that point right? So no sourcemaps for those?

**Comment by threepointone** [comment]
No, not generated till runtime. Can't generate sourcemaps except for very trivial cases.

**Comment by kentcdodds** [comment]
I still think that long class names would be kinda annoying :-/

**Comment by threepointone** [comment]
Noted, will keep in mind! This won't land immediately, will share progress and what I learn

**Comment by nicksrandall** [comment]
FWIW, my vote is `css-[label]-hash`.  

Do you plan to propagate this change to `glamor-Aphrodite` as well?

**Comment by dzearing** [comment]
Is this change in the works? Wondering if it's in progress or it already exists. I would definitely agree with comments above that `css-[label]-[hash]` would be great; or even just `[label]-[hash]`, no need for the css prefix.

**Comment by kentcdodds** [comment]
I would ask that some identifying thing is still available. [Glamorous relies](https://github.com/paypal/glamorous/blob/f411cf2a45095ef6396572d218b73fb016bb0783/src/get-glamor-classname.js#L15) on the `css-` prefix to not have to look up _every_ class name in glamor's `styleSheet`. Would be a shame to lose that optimization :)

**Comment by threepointone** [comment]
in the works, life getting in the way. the `css-`prefix should remain.

**Comment by knittingcodemonkey** [comment]
`css-customstyle_color-red_hover-color-blue_fontsize-20`

Are these directly from glamor, or would it make sense to use the JS css naming convention? 

`css-customstyle_fontSize-20`

The modifiers does present a different problem. Do we handle them all the same? Would it be 'hover-color-#f5f5f5' and 'css-before-customstyle_'?

**Comment by threepointone** [comment]
I'll write a longer comment in an hour or so with details and learnings. Hang tight my peeps!

**Comment by threepointone** [comment]
(In the next 24 hours 😔)

**Comment by kentcdodds** [comment]
No stress man. You've got lots going on!

**Comment by threepointone** [comment]
I've published v2.20.32 with an implementation of the above. basically css-[label]-[hash], as discussed. give it a spin if you're interested! I'll leave this issue open for a day or so while I try to break it, will close soon after that.

**Comment by threepointone** [comment]
seems alright, closing this, feel feel to reopen/open a new issue if you find a problem.
